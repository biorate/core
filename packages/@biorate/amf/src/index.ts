// Internal buffer for encoding
let tempBuffer = Buffer.allocUnsafe(1024);

// Internal buffer current offset
let tempPos = 0;

// Index for string reference encoding
let encodeCounter = 0;

// Storage for string reference encoding
let encodeCache: Record<string, number> = {};

// Storage for string reference decoding
let decodeCache: string[] = [];

/**
 * @description Resize buffer twice
 * @param {number} byte
 */
function resize(byte: number) {
  tempBuffer = Buffer.concat([tempBuffer, Buffer.allocUnsafe(byte)]);
}

/**
 * @description Write data to buffer
 * @param {*} data - Data to serialize and write
 */
function write(data: any) {
  tempPos = 0;
  encodeCounter = 0;
  encodeCache = {};
  writeValue(data);
  return tempBuffer.slice(0, tempPos);
}

/**
 * @description Check type of data and write recursive
 * @param {*} data - Data to serialize and write
 */
function writeValue(data: any) {
  if (data === undefined) return writeByte(0);
  if (data === null) return writeByte(1);
  if (data === false) return writeByte(2);
  if (data === true) return writeByte(3);

  var type = typeof data;

  if (type === 'number') {
    if (data >= 0 && data < 0x10000000 && data % 1 === 0) writeInt(data);
    else writeFloat(data);
  } else if (type === 'string') {
    writeString(data);
  } else if (Array.isArray(data)) {
    writeArray(data);
  } else if (type === 'object') {
    if (data instanceof Date) writeDate(data);
    else writeObject(data);
  } else {
    throw new Error(`${type} types not supported. Please check input data.`);
  }
}

/**
 * @description Write Date to internal buffer
 * @param {Number} value - Value to write
 */
function writeDate(value: Date) {
  writeByte(8);
  writeFloat(value.getTime(), false);
}

/**
 * @description Write byte to internal buffer
 * @param {Number} value - Value to write
 */
function writeByte(value: number) {
  if (tempBuffer.length <= tempPos) resize(1);
  tempBuffer.writeUInt8(value, tempPos);
  tempPos += 1;
}

/**
 * @description Write float64 to internal buffer
 * @param {Number} value - Value to write
 * @param {Boolean} [marker=true] - False if float marker should not be written
 */
function writeFloat(value: number, marker = true) {
  if (marker !== false) writeByte(5);
  if (tempBuffer.length <= tempPos + 8) resize(8);
  tempBuffer.writeDoubleBE(value, tempPos);
  tempPos += 8;
}

/**
 * @description Write uInt29 to internal buffer
 * @param {Number} value - Value to write
 * @param {Boolean} [marker=true] - False if int marker should not be written
 */
function writeInt(value: number, marker = true) {
  if (marker !== false) writeByte(4);
  if (value < 0x80) {
    writeByte(value);
    return;
  }
  if (value < 0x4000) {
    writeByte(0x80 | ((value >> 7) & 0x7f));
    writeByte(value & 0x7f);
    return;
  }
  if (value < 0x200000) {
    writeByte(0x80 | ((value >> 14) & 0x7f));
    writeByte(0x80 | ((value >> 7) & 0x7f));
    writeByte(value & 0x7f);
    return;
  }
  writeByte(0x80 | ((value >> 22) & 0x7f));
  writeByte(0x80 | ((value >> 15) & 0x7f));
  writeByte(0x80 | ((value >> 8) & 0x7f));
  writeByte(value & 0xff);
}

/**
 * @description Write string to internal buffer
 * @param {String} value - String to write
 * @param {Boolean} [marker=true] - False if int marker should not be wrote
 */
function writeString(value: string, marker = true) {
  let index = encodeCache[value];

  if (marker !== false) writeByte(6);

  if (index !== undefined) {
    writeInt(index << 1, false);
    return;
  }

  index = Buffer.byteLength(value);
  writeInt((index << 1) + 1, false);

  if (index) {
    encodeCache[value] = encodeCounter;
    ++encodeCounter;
    if (tempBuffer.length <= tempPos + index) resize(index);
    tempBuffer.write(value, tempPos);
    tempPos += index;
  }
}

/**
 * @description Write array to internal buffer
 * @param {Array} value - Array to write
 */
function writeArray(value: unknown[]) {
  let len = value.length;

  writeByte(9);
  writeInt((len << 1) + 1, false);
  writeString('', false);

  for (let i = 0; i < len; ++i) {
    // Should we check for function type?
    // We write size of array before, and should add one more loop for check
    // It will be expensive
    writeValue(value[i]);
  }
}

/**
 * @description Write object to internal buffer
 * @param {Object} value - Object to write
 */
function writeObject(value: Record<string, unknown>) {
  writeByte(10);
  writeInt(11, false);
  writeString('', false);

  let keys = Object.keys(value);

  for (let i = 0, l = keys.length; i < l; ++i) {
    let name = keys[i];
    let item = value[name];

    // For consistency, we should remove this check
    // But objects frequently have methods
    if (typeof item === 'function')
      throw new Error(`Function types not supported. Please check input data.`);

    writeString(name, false);
    writeValue(item);
  }

  writeString('', false);
}

/**
 * @description Deserialize AMF3 byte buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function read<T = any>(buffer: Buffer): T {
  tempPos = 0;
  decodeCache.length = 0;

  return readValue(buffer);
}

/**
 * @description Check flag type and read unserialized value
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readValue(buffer: Buffer): any {
  switch (readByte(buffer)) {
    case 0:
      return undefined;
    case 1:
      return null;
    case 2:
      return false;
    case 3:
      return true;
    case 4:
      return readInt(buffer);
    case 5:
      return readFloat(buffer);
    case 6:
      return readString(buffer);
    case 8:
      return readDate(buffer);
    case 9:
      return readArray(buffer);
    case 10:
      return readObject(buffer);
  }
}

/**
 * @description Read byte from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readByte(buffer: Buffer) {
  tempPos += 1;
  return buffer.readUInt8(tempPos - 1);
}

/**
 * @description Read Date object from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readDate(buffer: Buffer) {
  tempPos += 8;
  return new Date(buffer.readDoubleBE(tempPos - 8));
}

/**
 * @description Read float64 from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readFloat(buffer: Buffer) {
  tempPos += 8;
  return buffer.readDoubleBE(tempPos - 8);
}

/**
 * @description Read int29 from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readInt(buffer: Buffer) {
  let value: number;
  let byte: number;

  byte = readByte(buffer);
  if (byte < 0x80) return byte;

  value = (byte & 0x7f) << 7;
  byte = readByte(buffer);
  if (byte < 0x80) return value | byte;

  value = (value | (byte & 0x7f)) << 7;
  byte = readByte(buffer);
  if (byte < 0x80) return value | byte;

  value = (value | (byte & 0x7f)) << 8;
  byte = readByte(buffer);
  return value | byte;
}

/**
 * @description Read string from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readString(buffer: Buffer) {
  let len = readInt(buffer);

  if ((len & 1) === 0) return decodeCache[len >> 1];

  len = len >> 1;

  let string = '';

  if (len) {
    string = buffer.toString('utf8', tempPos, tempPos + len);
    decodeCache.push(string);
    tempPos += len;
  }

  return string;
}

/**
 * @description Read array from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readArray(buffer: Buffer) {
  let len = readInt(buffer);
  len = len >> 1;

  readString(buffer);

  let result = [];

  for (let i = 0; i < len; ++i) result.push(readValue(buffer));

  return result;
}

/**
 * @description Read object from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readObject(buffer: Buffer) {
  let name;
  readInt(buffer);
  readString(buffer);

  let result: Record<string, unknown> = {};

  while ((name = readString(buffer))) result[name] = readValue(buffer);

  return result;
}

/**
 * @description AMF3 encode function
 *
 * Links:
 * - [wikipedia](https://en.wikipedia.org/wiki/Action_Message_Format)
 * - [specification] https://www.adobe.com/content/dam/acom/en/devnet/pdf/amf-file-format-spec.pdf
 *
 * Limitations:
 * - Function is not supported
 * - XML is not supported
 * - ByteArray is not supported
 * - Dictionary is not supported
 * - Vectors is not supported
 *
 * @example
 * ```ts
 * import { encode } from '@biorate/amf';
 *
 * const buffer = encode({ test: 1 });
 *
 * console.log(buffer); // <Buffer 0a 0b 01 09 74 65 73 74 04 01 01>
 * ```
 */
export const encode = write;

/**
 * @description AMF3 decode function
 *
 * Limitations:
 *   - see encode description section
 *
 * @example
 * ```ts
 * import { encode, decode } from '@biorate/amf';
 *
 * const buffer = encode({ test: 1 });
 *
 * console.log(decode(buffer)); // { test: 1 }
 * ```
 */
export const decode = read;
