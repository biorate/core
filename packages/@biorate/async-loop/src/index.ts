import { timer } from '@biorate/tools';

/**
 * @description Async loop implementation
 *
 * @example
 * ```
 * import { AsyncLoop } from '@biorate/async-loop';
 *
 * (async () => {
 *   let i = 0;
 *   const loop = new AsyncLoop(
 *     () => {
 *       console.log(++i); // 1, 2, 3 ... e.t.c
 *     },
 *     console.error,
 *     1000,
 *   );
 * })();
 * ```
 */
export class AsyncLoop {
  #process: () => Promise<void> | void;

  #error: (e: unknown) => Promise<void> | void;

  #interval: number;

  #paused = false;

  #stoped = false;

  public constructor(
    process: () => Promise<void> | void,
    error?: () => Promise<void> | void,
    interval = 1000,
  ) {
    this.#process = process;
    this.#error = error ?? console.error;
    this.#interval = interval;
    this.process().catch(this.#error);
  }

  public pause() {
    this.#paused = true;
  }

  public resume() {
    this.#paused = false;
  }

  public stop() {
    this.#stoped = true;
  }

  protected async process() {
    while (true) {
      await timer.wait(this.#interval);
      if (this.#paused) continue;
      if (this.#stoped) break;
      try {
        await this.#process();
      } catch (e: unknown) {
        await this.#error(e);
      }
    }
  }
}
