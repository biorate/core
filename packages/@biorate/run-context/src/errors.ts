import { BaseError } from '@biorate/errors';

export class RunContextKeyAlreadyExistsError extends BaseError {
  public constructor() {
    super(`Key already exists`);
  }
}
