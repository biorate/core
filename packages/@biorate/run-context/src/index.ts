import 'reflect-metadata';
export * from './context';
export * from './scenario';
