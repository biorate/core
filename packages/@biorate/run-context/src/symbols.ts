import { create } from '@biorate/symbolic';

const Symbols = create('Context');

export const Scenario = Symbols.Scenario;
