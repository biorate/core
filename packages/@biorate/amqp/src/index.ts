import { EventEmitter } from 'events';
import { injectable } from '@biorate/inversion';
import { Connector } from '@biorate/connector';
import { events } from '@biorate/tools';
import { connect, ChannelWrapper } from 'amqp-connection-manager';
import { AmqpCantConnectError, ChannelNotExistsError } from './errors';
import { IAmqpConfig, IAmqpConnection, ICreateChannelOpts } from './interfaces';

export { ConsumeMessage } from 'amqplib/properties';
export * from 'amqp-connection-manager';
export * from './errors';
export * from './interfaces';

/**
 * @description Amqp connector
 *
 * ### Features:
 * - connection manager for AMQP
 *
 * @example
 * ```
 * import { inject, container, Types, Core } from '@biorate/inversion';
 * import { IConfig, Config } from '@biorate/config';
 * import { AmqpConnector } from '@biorate/amqp';
 *
 * const connectionName = 'amqp';
 * const channelName = 'test';
 *
 * export class Root extends Core() {
 *   @inject(AmqpConnector) public connector: AmqpConnector;
 * }
 *
 * container.bind<IConfig>(Types.Config).to(Config).inSingletonScope();
 * container.bind<AmqpConnector>(AmqpConnector).toSelf().inSingletonScope();
 * container.bind<Root>(Root).toSelf().inSingletonScope();
 *
 * container.get<IConfig>(Types.Config).merge({
 *   Amqp: [
 *     {
 *       name: connectionName,
 *       urls: ['amqp://localhost:5672'],
 *     },
 *   ],
 * });
 *
 * const root = <Root>container.get<Root>(Root);
 *
 * (async () => {
 *   root.connector.createChannel(connectionName, {
 *     name: channelName,
 *     json: true,
 *     setup: async (channel: Channel) => {
 *       await channel.assertExchange('test-exchange', 'topic');
 *       await channel.assertQueue('test-queue', { exclusive: true, autoDelete: true });
 *       await channel.bindQueue('test-queue', 'test-exchange', '#send');
 *       await channel.consume('test-queue', (data: ConsumeMessage | null) => {
 *         console.log(data?.content?.toString?.()); // {"test": 1}
 *       });
 *     },
 *   });
 *   root.connector.channel(channelName)!.publish('test-exchange', '#send', { test: 1 });
 * })();
 * ```
 */
@injectable()
export class AmqpConnector extends Connector<IAmqpConfig, IAmqpConnection> {
  /**
   * @description Private connections storage
   */
  private '#connections': Map<string, IAmqpConnection>;
  /**
   * @description Private link to selected (used) connection
   */
  private '#current': IAmqpConnection | undefined;
  /**
   * @description Namespace path for fetching configuration
   */
  protected readonly namespace = 'Amqp';
  /**
   * @description Channels storage
   */
  protected readonly channels = new Map<string, ChannelWrapper>();
  /**
   * @description Create connection
   */
  protected async connect(config: IAmqpConfig) {
    let connection: IAmqpConnection;
    try {
      connection = connect(config.urls, config.options);
      await events.once(<EventEmitter>(<unknown>connection), 'connect');
      connection.on('disconnect', console.warn);
    } catch (e: unknown) {
      throw new AmqpCantConnectError(<Error>e);
    }
    return connection;
  }
  /**
   * @description Create channel method
   */
  public createChannel(name: string, options: ICreateChannelOpts) {
    const channel = this.connection(name).createChannel(options);
    this.channels.set(options.name, channel);
    return channel;
  }
  /**
   * @description Get channel method
   */
  public channel(name: string) {
    if (!this.channels.has(name)) throw new ChannelNotExistsError(name);
    return this.channels.get(name);
  }
}
