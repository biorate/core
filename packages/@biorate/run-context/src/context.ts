import { object as o } from '@biorate/tools';
import { Scenario as ScenarioSymbol } from './symbols';
import { RunContextKeyAlreadyExistsError } from './errors';
/**
 * @description Create context and run scenarios
 *
 * @example
 * ```
 * import { Scenario, Context, step } from '@biorate/run-context';
 *
 * export class Scenario1 extends Scenario {
 *   @step()
 *   protected async step1() {
 *     this.ctx.set('step1', true);
 *   }
 *
 *   @step()
 *   protected async step2() {
 *     this.ctx.set('step2', 1);
 *   }
 * }
 *
 * export class Scenario2 extends Scenario {
 *   @step()
 *   protected async step3() {
 *     this.ctx.set('step3', false);
 *   }
 * }
 *
 * (async () => {
 *   const ctx = await Context.run([Scenario1, Scenario2], { initial: 'value' });
 *   console.log(ctx.get<boolean>('step1')); // true
 *   console.log(ctx.get<number>('step2')); // 1
 *   console.log(ctx.get<boolean>('step3')); // false
 *   console.log(ctx.get()); // {
 *   //    'step1': true,
 *   //    'step2': 1,
 *   //    'step3': false
 *   //  }
 * })();
 * ```
 */
export class Context {
  public static async run(scenarios: any[], ctx: Record<string, any> = {}) {
    const instance = new this(ctx);
    for (const Scenario of scenarios) {
      const scenario = new Scenario(instance);
      const methods = new Set();
      const steps: (() => void | Promise<void>)[] = [];
      const objects: any[] = [];
      o.walkProto(scenario, (object: any) => objects.unshift(object));
      for (const object of objects) {
        const names = Object.getOwnPropertyNames(object);
        for (const name of names) {
          if (methods.has(name)) continue;
          methods.add(name);
          const descriptor = Object.getOwnPropertyDescriptor(object, name);
          if (!descriptor?.value) continue;
          const meta = Reflect.getMetadata(this.metaKey, descriptor.value);
          if (!meta) continue;
          const step = scenario[name].bind(scenario);
          Reflect.defineMetadata(this.metaKey, { name: meta.name }, step);
          steps.push(step);
        }
      }
      for (const step of steps) await this.runStep(step);
    }
    return instance;
  }

  protected static metaKey = ScenarioSymbol;

  protected static async runStep(step: () => void | Promise<void>) {
    await step();
  }

  #ctx: Record<string, any> = {};

  protected constructor(ctx: Record<string, any>) {
    for (const key in ctx) this.set(key, ctx[key]);
  }

  public set(key: string, value: any) {
    if ('key' in this.#ctx) throw new RunContextKeyAlreadyExistsError();
    this.#ctx[key] = value;
  }

  public get<T = unknown>(key?: string) {
    return key ? <T>this.#ctx[key] : this.#ctx;
  }

  public del(key: string) {
    return delete this.#ctx[key];
  }
}
