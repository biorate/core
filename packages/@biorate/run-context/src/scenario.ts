import { Context } from './context';
import { Scenario as ScenarioSymbol } from './symbols';
/**
 * @description Abstract Scenario class
 */
export abstract class Scenario {
  public constructor(protected ctx: Context) {}
}

export const step =
  (name?: string) => (target: any, propertyKey: string, descriptor: PropertyDescriptor) =>
    Reflect.defineMetadata(ScenarioSymbol, { name }, descriptor?.value);
