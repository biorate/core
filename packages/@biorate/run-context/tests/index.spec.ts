import { expect } from 'chai';
import { Scenario1, Scenario2 } from './__mocks__';
import { Context } from '../src';

describe('@biorate/run-context', function () {
  it('Context.run', async () => {
    const ctx = await Context.run([Scenario1, Scenario2], { initial: 'value' });
    expect(ctx.get()).toMatchSnapshot();
  });
});
