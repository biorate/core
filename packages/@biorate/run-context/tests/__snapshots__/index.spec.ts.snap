// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@biorate/run-context Context.run 1`] = `
Object {
  "initial": "value",
  "step0-Scenario1": true,
  "step1-Scenario1": true,
  "step1-Scenario2": true,
  "step2-Scenario1": true,
  "step2-Scenario2": true,
}
`;
